Fix and refactor `deprecated` decorator to `deprecate_func`

lagru opened the pull request:
## Description

- Closes #6585. `skimage_deprecation` was only used in the decorator `deprecated` while everywhere else FutureWarning is used. So let's be consistent about this. 
- Also don't modify the warning filter when calling the wrapped function. That this happens is very implicit and hidden from the user. I'd argue that it's the users responsibility to configure his warning filter while it's our responsibility to chose good and expected defaults.
- Closes #6586. Previously, the warning did not point to where the deprecated function was called (user code) but at the deprecated function's definition (library code)! Add a test for this.
- I also propose an updated signature of the decorator, that makes sure that the deprecated version is specified (forgotten in previous uses of this decorator). I replaced `alt_func` with a more general `hint` parameter as there is not always a clear alternative function. In that case it's possible to provide a more flexible guiding message to the user. It also uses the imperative mood like other similar decorators.
- Also do away with the behavior parameter. Raising an exception defeats the purpose of a deprecation decorator as it immediately breaks user code. It impossible to use the deprecated function, in which case it should just be removed.
- Update the current uses of the decorator.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

stefanv commented:
Thanks for the refactor, @lagru.

I wonder if it's worth capturing the version in which a function was deprecated; does that matter much to the user? We didn't have that before.

lagru commented:
Yes, I think it's useful to provide the warning when something was deprecated. It makes it easier to find related resources (e.g. release notes) and increases the specificity of the search message. That's useful to users and to us during user support and maintenance as well. It might even prevent errors such as the on reverted in https://github.com/scikit-image/scikit-image/pull/6583/commits/998ee00833194a3fab29d0816737a309d382ab84. Also, this pattern is already used in our [`deprecate_kwarg` decorator](https://github.com/scikit-image/scikit-image/blob/e0e7ed7fecf33986553c9b6ca362a36bd9c4fcc1/skimage/_shared/utils.py#L237-L238).

stefanv commented:
OK, in that case I'd use the wording "X was deprecated in skimage v0.19, and will be removed in v0.22"

mkcor commented:
> "X was deprecated in skimage v0.19, and will be removed in v0.22"

Just to be on the same page, shouldn't it be "X was deprecated in skimage v0.19, and will be removed in **v0.21**" in the general case (deprecation cycle of two minor releases)?

mkcor commented on skimage/_shared/tests/test_utils.py:
```diff
@@ -300,3 +300,27 @@ def test_decorator_warnings():
     assert record[0].filename == __file__
+
+
+@deprecate_func(
```
```suggestion
@deprecated_func(
```
I'm not sure about my suggestion: I think that *formally* it helps with consistency (`deprecated_func` replaced `deprecated`, plus the dummy function below is named `_deprecated_func`), but maybe you intended that and you are absolutely logical (the decorator *deprecates* while the dummy function stands for a *deprecated* function).

grlee77 replied:
It seems okay to leave as `deprecate_func`, but no strong opinion either way

mkcor replied:
@grlee77 ok, thanks.
@lagru feel free to ignore this suggestion.

mkcor commented on skimage/_shared/tests/test_utils.py:
```diff
@@ -300,3 +300,27 @@ def test_decorator_warnings():
+    """
+
+
+def test_deprecate_func():
```
```suggestion
def test_deprecated_func():
```

mkcor commented on skimage/_shared/utils.py:
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
-class deprecated:
-    """Decorator to mark deprecated functions with warning.
+class deprecate_func(_DecoratorBaseClass):
+    """Decorator a deprecated functions and warn when it is called.
```
```suggestion
    """Decorate a deprecated function and warn when it is called.
```

mkcor commented on skimage/_shared/utils.py:
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
     removed_version : str
         The package version in which the deprecated function will be removed.
+    hint : str, optional
+        A hint on how to address this deprecation.
+        E.g. "Use `skimage.submodule.alternative_func` instead."
```
```suggestion
        A hint on how to address this deprecation,
        e.g., "Use `skimage.submodule.alternative_func` instead."
```

mkcor commented on skimage/_shared/utils.py:
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
-        Behavior during call to deprecated function: 'warn' = warn user that
-        function is deprecated; 'raise' = raise error.
+    deprecated_version : str
+        The package version during which the deprecation was introduced.
```
```suggestion
        The package version when the deprecation was introduced.
```

mkcor commented on skimage/feature/__init__.py:
```diff
@@ -30,15 +30,17 @@
-@deprecated(alt_func='skimage.feature.graycomatrix',
-            removed_version='1.0')
+@deprecate_func(hint='Use `skimage.feature.graycomatrix` instead.',
+                deprecated_version="0.19",
```
```suggestion
                deprecated_version='0.19',
```

mkcor commented on skimage/feature/__init__.py:
```diff
@@ -30,15 +30,17 @@
-@deprecated(alt_func='skimage.feature.graycoprops',
-            removed_version='1.0')
+@deprecate_func(hint='Use `skimage.feature.graycoprops` instead.',
+                deprecated_version="0.19",
```
```suggestion
                deprecated_version='0.19',
```

mkcor commented on skimage/_shared/utils.py:
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
+            )
             return func(*args, **kwargs)
 
         # modify doc string to display deprecation warning
```
```suggestion
        # modify docstring to display deprecation warning
```

grlee77 commented on skimage/_shared/utils.py:
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
+            f"`{func.__name__}` is deprecated since version "
+            f"{self.deprecated_version}"
+        )
+        if self.deprecated_version:
```
typo?
```suggestion
        if self.removed_version:
```

mkcor replied:
Looks like it, good catch!

grlee77 reviewed (commented):
There are some comments and conflicts to resolve, but it looks good to me once that is done

lagru commented:
Thanks for the reviews both of you! I addressed your comments and resolved the conflicts. I think the [CI failure](https://github.com/scikit-image/scikit-image/actions/runs/3637875098/jobs/6139398577#step:6:623) is unrelated so this should be good to go from my side.

lagru commented:
Maybe we can still get this bug fix into 0.20. :pray:

stefanv commented:
This is fine with me, if you want to pick it back up.

jarrodmillman commented on skimage/filters/ridges.py:
```diff
@@ -14,12 +14,17 @@ class of ridge filters relies on the eigenvalues of the Hessian matrix of
-@deprecated(alt_func='skimage.feature.hessian_matrix', removed_version="0.21")
+@deprecate_func(
+    deprecated_version="0.20",
+    removed_version="0.21",
```
```suggestion
    removed_version="0.22",
```

jarrodmillman replied:
Also we should add `compute_hessian_eigenvalues` to `TODO.txt`.

jarrodmillman reviewed (approved):
LGTM (w/ the suggested changes).

stefanv commented:
Thanks, Lars!
